File: playwright-python/pages/checkout_page.py
import re
from decimal import Decimal

from playwright.sync_api import Browser, Page, expect


class CheckoutPage:
    FIELD_IDS = {
        "Full name": "#id_full_name",
        "Email": "#id_email",
        "Phone": "#id_phone",
        "Address": "#id_address",
        "Postal code": "#id_postal_code",
        "City": "#id_city",
        "Country": "#id_country",
    }

    def __init__(self, page: Page, base_url: str):
        self.page = page
        self.base_url = base_url.rstrip("/")
        self.order_id: int | None = None

    def open(self) -> None:
        self.page.goto(f"{self.base_url}/checkout")

    def expect_storefront(self) -> None:
        expect(self.page).to_have_url(f"{self.base_url}/")

    def expect_shipping_page(self) -> None:
        expect(self.page).to_have_url(f"{self.base_url}/checkout")
        expect(self.page.get_by_role("heading", name="Shipping details")).to_be_visible()

    def expect_shipping_fields(self) -> None:
        for label in ("Full name", "Email", "Phone", "Address", "Postal code", "City", "Country"):
            expect(self.page.get_by_label(label, exact=True)).to_be_visible()

    def expect_summary(self, product_name: str, quantity: int, total: Decimal) -> None:
        summary = self.page.locator(".checkout-summary")
        expect(summary).to_contain_text(product_name)
        expect(summary).to_contain_text(f"{quantity} ")
        expect(summary).to_contain_text(f"${total:.2f}")

    def back_to_cart(self) -> None:
        self.page.get_by_role("link", name="Back to cart", exact=True).click()

    def fill_shipping(self, data: dict[str, str], missing: str | None = None) -> None:
        for label, value in data.items():
            self.page.get_by_label(label, exact=True).fill("" if label == missing else value)

    def place_order(self) -> None:
        self.page.get_by_role("button", name="Place order", exact=True).click()

    def expect_field_validation(self, label: str) -> None:
        assert self.page.url == f"{self.base_url}/checkout", (
            f"Expected checkout validation page, got {self.page.url}"
        )
        field = self.page.locator(self.FIELD_IDS[label])
        expect(field).to_be_visible()
        assert not field.evaluate("element => element.validity.valid")
        assert field.evaluate("element => element.validationMessage")

    def expect_checkout_url(self) -> None:
        expect(self.page).to_have_url(f"{self.base_url}/checkout")

    def expect_confirmation(self) -> None:
        expect(self.page).to_have_url(re.compile(r"/orders/\d+/confirmation$"))
        match = re.search(r"/orders/(\d+)/confirmation$", self.page.url)
        assert match
        self.order_id = int(match.group(1))
        expect(
            self.page.get_by_role(
                "heading", name="Your order was placed successfully", exact=True
            )
        ).to_be_visible()

    def expect_order_number(self) -> None:
        expect(self.page.locator(".confirmation > .eyebrow")).to_have_text(
            re.compile(r"Order #\d+")
        )

    def expect_success_message(self) -> None:
        expect(
            self.page.get_by_role(
                "heading", name="Your order was placed successfully", exact=True
            )
        ).to_be_visible()

    def expect_order_items(self, expected: list[dict[str, object]]) -> None:
        for item in expected:
            line = self.page.locator(".confirmation .summary-lines li").filter(
                has_text=str(item["name"])
            )
            expect(line).to_contain_text(str(item["quantity"]))
            expect(line).to_contain_text(f'${item["unit_price"]:.2f}')
            expect(line).to_contain_text(f'${item["line_total"]:.2f}')

    def expect_order_total(self, total: Decimal) -> None:
        expect(self.page.locator(".confirmation .summary-total")).to_contain_text(
            f"${total:.2f}"
        )

    def expect_shipping_summary(self, data: dict[str, str]) -> None:
        summary = self.page.locator('[aria-labelledby="delivery-title"]')
        for value in data.values():
            expect(summary).to_contain_text(value)

    def back_to_storefront(self) -> None:
        self.page.get_by_role("link", name="Back to storefront", exact=True).click()

    def expect_storefront_catalog(self) -> None:
        expect(self.page.get_by_role("heading", name="Welcome to FictoShop")).to_be_visible()
        expect(self.page.get_by_role("heading", name="Our products")).to_be_visible()

    def expect_insufficient_stock(self, product_name: str) -> None:
        expect(self.page.locator(".form-error")).to_contain_text(
            f"There is no longer enough stock for {product_name}."
        )

    def confirmation_denied_in_another_session(self, browser: Browser) -> tuple[int, str]:
        context = browser.new_context()
        try:
            response = context.new_page().goto(self.page.url)
            assert response is not None
            return response.status, response.text()
        finally:
            context.close()

File: playwright-typescript/src/pages/checkout-page.ts
import assert from "node:assert/strict";
import { expect, type Browser, type Page } from "@playwright/test";
import type { OrderItemRecord } from "../support/models.js";
import { settings } from "../support/settings.js";

const fieldIds: Record<string, string> = {
  "Full name": "#id_full_name",
  Email: "#id_email",
  Phone: "#id_phone",
  Address: "#id_address",
  "Postal code": "#id_postal_code",
  City: "#id_city",
  Country: "#id_country"
};

export class CheckoutPage {
  orderId?: number;

  constructor(private readonly page: Page) {}

  open(): Promise<unknown> { return this.page.goto(`${settings.baseUrl}/checkout`); }
  expectStorefront(): Promise<void> { return expect(this.page).toHaveURL(`${settings.baseUrl}/`); }

  async expectShippingPage(): Promise<void> {
    await expect(this.page).toHaveURL(`${settings.baseUrl}/checkout`);
    await expect(this.page.getByRole("heading", { name: "Shipping details" })).toBeVisible();
  }

  async expectShippingFields(): Promise<void> {
    for (const label of Object.keys(fieldIds)) {
      await expect(this.page.getByLabel(label, { exact: true })).toBeVisible();
    }
  }

  async expectSummary(name: string, quantity: number, total: number): Promise<void> {
    const summary = this.page.locator(".checkout-summary");
    await expect(summary).toContainText(name);
    await expect(summary).toContainText(`${quantity} `);
    await expect(summary).toContainText(`$${total.toFixed(2)}`);
  }

  backToCart(): Promise<void> { return this.page.getByRole("link", { name: "Back to cart", exact: true }).click(); }

  async fillShipping(data: Record<string, string>, missing?: string): Promise<void> {
    for (const [label, value] of Object.entries(data)) {
      await this.page.getByLabel(label, { exact: true }).fill(label === missing ? "" : value);
    }
  }

  placeOrder(): Promise<void> { return this.page.getByRole("button", { name: "Place order", exact: true }).click(); }

  async expectFieldValidation(label: string): Promise<void> {
    assert.equal(this.page.url(), `${settings.baseUrl}/checkout`);
    const field = this.page.locator(fieldIds[label]);
    await expect(field).toBeVisible();
    assert.equal(await field.evaluate((element: HTMLInputElement) => element.validity.valid), false);
    assert.ok(await field.evaluate((element: HTMLInputElement) => element.validationMessage));
  }

  expectCheckoutUrl(): Promise<void> { return expect(this.page).toHaveURL(`${settings.baseUrl}/checkout`); }

  async expectConfirmation(): Promise<void> {
    await expect(this.page).toHaveURL(/\/orders\/\d+\/confirmation$/);
    const match = this.page.url().match(/\/orders\/(\d+)\/confirmation$/);
    assert.ok(match);
    this.orderId = Number(match[1]);
    await expect(this.page.getByRole("heading", { name: "Your order was placed successfully", exact: true })).toBeVisible();
  }

  expectOrderNumber(): Promise<void> { return expect(this.page.locator(".confirmation > .eyebrow")).toHaveText(/Order #\d+/); }
  expectSuccess(): Promise<void> { return expect(this.page.getByRole("heading", { name: "Your order was placed successfully", exact: true })).toBeVisible(); }

  async expectOrderItems(expected: OrderItemRecord[]): Promise<void> {
    for (const item of expected) {
      const line = this.page.locator(".confirmation .summary-lines li").filter({ hasText: item.product_name });
      await expect(line).toContainText(String(item.quantity));
      await expect(line).toContainText(`$${item.unit_price.toFixed(2)}`);
      await expect(line).toContainText(`$${item.line_total.toFixed(2)}`);
    }
  }

  expectOrderTotal(total: number): Promise<void> { return expect(this.page.locator(".confirmation .summary-total")).toContainText(`$${total.toFixed(2)}`); }

  async expectShippingSummary(data: Record<string, string>): Promise<void> {
    const summary = this.page.locator('[aria-labelledby="delivery-title"]');
    for (const value of Object.values(data)) {
      await expect(summary).toContainText(value);
    }
  }

  backToStorefront(): Promise<void> { return this.page.getByRole("link", { name: "Back to storefront", exact: true }).click(); }

  async expectStorefrontCatalog(): Promise<void> {
    await expect(this.page.getByRole("heading", { name: "Welcome to FictoShop" })).toBeVisible();
    await expect(this.page.getByRole("heading", { name: "Our products" })).toBeVisible();
  }

  expectInsufficientStock(name: string): Promise<void> {
    return expect(this.page.locator(".form-error")).toContainText(`There is no longer enough stock for ${name}.`);
  }

  async openConfirmationInAnotherSession(browser: Browser): Promise<{ status: number; body: string }> {
    const context = await browser.newContext();
    try {
      const response = await (await context.newPage()).goto(this.page.url());
      assert.ok(response);
      return { status: response.status(), body: await response.text() };
    } finally {
      await context.close();
    }
  }
}
